src/frontend/js/pages/TeacherDashboardContractsLayout/hooks/useDownloadContractArchive/index.tsx

```diff
@@ -0,0 +1,107 @@
+import { useEffect, useState } from 'react';
+import useContractArchive from 'hooks/useContractArchive';
+import { useContracts } from 'hooks/useContracts';
+import { ContractState } from 'types/Joanie';
+import { PER_PAGE } from 'settings';
+import useCheckContractArchiveExist from '../useCheckContractArchiveExist';
+import {
+  storeContractArchiveId,
+  unstoreContractArchiveId,
+  getStoredContractArchiveId,
+  isStoredContractArchiveIdExpired,
+} from './contractArchiveLocalStorage';
+
+export enum ContractDownloadStatus {
+  IDDLE = 'iddle',
+  PENDING = 'pending',
+  READY = 'ready',
+}
+
+interface UseTeacherContractsBulkDownloadProps {
+  organizationId: string;
+}
+
+const useDownloadContractArchive = ({ organizationId }: UseTeacherContractsBulkDownloadProps) => {
+  const { items: contracts } = useContracts({
+    organization_id: organizationId,
+    signature_state: ContractState.SIGNED,
+    page: 1,
+    page_size: PER_PAGE.teacherContractList,
+  });
+  const hasContractToDownload = contracts.length > 0;
+
+  const {
+    methods: { get: getArchive, create: createArchive },
+  } = useContractArchive();
+
+  const [contractArchiveId, setContractArchiveId] = useState<string | null>(
+    getStoredContractArchiveId(),
+  );
+  const { isContractArchiveExist, doCheckArchiveExist } = useCheckContractArchiveExist({
+    contractArchiveId,
+  });
+
+  const [isDownloadRequest, setIsDownloadRequest] = useState(false);
+  const [contractDownloadStatus, setContractDownloadStatus] = useState<ContractDownloadStatus>(
+    ContractDownloadStatus.IDDLE,
+  );
+
+  const onArchiveGenerationStart = (archiveId: string) => {
+    setIsDownloadRequest(true);
+    setContractArchiveId(archiveId);
+    storeContractArchiveId(archiveId);
+
+    doCheckArchiveExist(archiveId);
+  };
+
+  const onArchiveDownloadDone = () => {
+    setIsDownloadRequest(false);
+    setContractArchiveId(null);
+    unstoreContractArchiveId();
+  };
+
+  const doDownloadArchive = async () => {
+    if (isContractArchiveExist && contractArchiveId !== null) {
```

Comment:
`contractArchiveId !== null` is there for ts, i could replace it by `await getArchive(contractArchiveId!);` as when contractArchiveId is null, isContractArchiveExist is false.